File: mill-plugin/src/net/exoego/uika/mill/Uika.scala
package net.exoego.uika.mill

import coursier.core as cs
import mill.*
import mill.api.{Discover, Evaluator, ExternalModule, SelectMode}
import mill.javalib.{BoundDep, CoursierModule, Dep, JavaModule, TestModule}
import net.exoego.uika.plugin.core.{ClasspathDump, DumpFormat, JfrEvidence, UikaCli}

import scala.annotation.nowarn
import scala.jdk.CollectionConverters.*

/**
 * Mill entry points for uika, invoked as `./mill net.exoego.uika.mill.Uika/dumpClasspath`.
 *
 * An [[ExternalModule]] rather than a trait users mix in, so a build of any size is wired up by
 * the `//| mvnDeps` header alone: the commands find every non-test `JavaModule` through the
 * `Evaluator` themselves. The one thing this shape cannot reach is a test JVM's `forkArgs`, so
 * JFR class-load collection is the one part that does need a mixin ([[UikaTestModule]]).
 */
object Uika extends ExternalModule {

  /**
   * Writes every non-test module's resolved runtime classpath as a uika v2 dump.
   *
   * Evaluating a module's `localRunClasspath` compiles it, so the build outputs the dump points
   * at exist by the time the CLI scans them. That mirrors the sbt plugin; Mill has no
   * resolution-only mode to opt into because a Mill module cannot resolve its own runtime
   * classpath without its upstream modules' compile output existing anyway.
   */
  def dumpClasspath(ev: Evaluator, output: String = "") = Task.Command(exclusive = true) {
    val modules = javaModules(ev)
    if (modules.isEmpty) {
      Task.fail("uika: no JavaModule found in this build")
    }
    val dumps = ev.execute(modules.map(moduleDumpTask)).values.get
    val workspace = Task.ctx().workspace
    val out =
      if (output.isEmpty) workspace / "out" / "uika" / "classpath.json"
      else os.Path(output, workspace)
    os.makeDir.all(out / os.up)
    os.write.over(
      out,
      DumpFormat.writeV2(
        dumps.asJava,
        Seq(workspace.toString).asJava,
        DumpFormat.dumpRelease(dumps.asJava)
      )
    )
    Task.log.info(s"uika classpath dump: $out")
    out.toString
  }

  /**
   * Runs `uika upgrade-check` over a before/after pair of dumps, fetching the CLI itself as
   * `net.exoego.uika:uika-cli:<version>:<platform>@zip` through Mill's own resolution.
   *
   * @param jdkRelease resolve JDK hierarchy escapes against this API release; 0 disables the
   *                   layer and a negative value, the default, derives the lowest release any
   *                   module compiles for, else the build JVM's, clamped by
   *                   [[UikaCli.effectiveJdkRelease]] to what its ct.sym serves
   * @param jfr        a directory of JFR recordings from a test run of the current, not yet
   *                   upgraded build, or a single `.jfr` recording
   */
  def upgradeCheck(
      ev: Evaluator,
      before: String,
      after: String,
      failOn: String = "any",
      excludeFile: Seq[String] = Nil,
      jdkRelease: Int = -1,
      jfr: String = "",
      draftExcludeFile: String = "",
      cliVersion: String = ""
  // persistent so `Task.dest` survives: Mill wipes a non-persistent dest before every run,
  // which would defeat both UikaCli.extractBinary's skip-if-present and JfrEvidence.rewrite's
  // stale-conversion sweep. The other three plugins extract into their build directory.
  ) = Task.Command(exclusive = true, persistent = true) {
    val version = cliVersion match {
      case "" =>
        Option(getClass.getPackage.getImplementationVersion).filter(_.nonEmpty).getOrElse(
          Task.fail("uika-cli version is unknown; pass --cliVersion <version>")
        )
      case v => v
    }
    val workspace = Task.ctx().workspace
    val log: java.util.function.Consumer[String] = line => Task.log.info(line)
    // The CLI ZIP goes through a build module's own resolver, so custom `repositories`,
    // mirrors and credentials are the build's. Any module will do: repositories are declared
    // on a shared trait in practice. Failing rather than falling back to this ExternalModule's
    // own resolver keeps that promise -- a `defaultResolver()` call here would be lifted into
    // an unconditional task edge by the command macro and evaluated even on the Some branch.
    val modules = javaModules(ev)
    val resolver = modules.headOption match {
      case Some(m) => ev.execute(Seq(m.defaultResolver)).values.get.head
      case None => Task.fail("uika: no JavaModule found in this build")
    }
    val binary = extractCli(resolver, version, Task.dest)
    // Recordings are converted here, never handed to the CLI: the CLI is JVM-free and must
    // not read binary JFR.
    val classLoadLogs = JfrEvidence.rewrite(
      Option(jfr).filter(_.nonEmpty).map(os.Path(_, workspace).toNIO).toSeq.asJava,
      (Task.dest / JfrEvidence.WORK_DIR_NAME).toNIO,
      log
    )
    // The LOWEST release any module compiles for, because one flag serves a run that checks
    // every module. Under-claiming only costs Unknowns, while over-claiming makes a member
    // the runtime lacks resolve cleanly and loses the finding with nothing to show. A build
    // declaring nothing falls back to the JVM, the only evidence left. The dump keeps each
    // module's own release next to it (moduleDumpTask); the flag stays one value because the
    // layer it switches on is process-wide.
    //
    // mandatoryJavacOptions as well as javacOptions, since Mill compiles with both and a
    // trait that pins the release commonly does it there.
    val jdk = UikaCli.JdkSource.current()
    val wantedRelease =
      if (jdkRelease >= 0) jdkRelease
      else {
        val optionTasks = modules.map(_.javacOptions) ++ modules.map(_.mandatoryJavacOptions)
        val declared = ev.execute(optionTasks).values.get
          .flatMap(options => Option(UikaCli.declaredRelease(options.asJava)).map(_.intValue))
        if (declared.isEmpty) Runtime.version().feature() else declared.min
      }
    val exit = UikaCli.runUpgradeCheck(
      binary,
      os.Path(before, workspace).toNIO,
      os.Path(after, workspace).toNIO,
      failOn,
      excludeFile.map(os.Path(_, workspace).toNIO).asJava,
      UikaCli.effectiveJdkRelease(wantedRelease, jdk, log),
      jdk,
      classLoadLogs,
      Option(draftExcludeFile).filter(_.nonEmpty).map(os.Path(_, workspace).toNIO).orNull,
      log
    )
    exit match {
      case 0 => ()
      case 1 => Task.fail("uika upgrade-check found broken references (see output above)")
      case n => Task.fail(s"uika upgrade-check failed with exit code $n")
    }
  }

  /**
   * Extracts the platform's uika binary, resolving the ZIP through Mill's coursier setup so
   * mirrors, credentials and the cache are the build's own. `artifactTypes` has to name zip:
   * coursier's default set is jar-shaped and would drop the distribution entirely.
   */
  private def extractCli(
      resolver: CoursierModule.Resolver,
      version: String,
      dest: os.Path
  )(using mill.api.TaskCtx): java.nio.file.Path = {
    val classifier = UikaCli.platformClassifier()
    // Intransitive, as in all three sibling plugins: the distribution is a native binary, and
    // anything the POM ever gains would be downloaded and could win the zip pick below.
    val dep = Dep.parse(
      s"${UikaCli.GROUP}:${UikaCli.ARTIFACT}:$version;classifier=$classifier;type=zip"
    ).exclude("*" -> "*")
    val resolved =
      resolver.classpath(Seq(dep), artifactTypes = Some(Set(coursier.Type("zip")))).map(_.path)
    val zip = resolved.find(_.last.endsWith(".zip")).getOrElse(
      Task.fail(s"uika-cli zip not found among ${resolved.mkString(", ")}")
    )
    UikaCli.extractBinary(zip.toNIO, (dest / s"cli-$version-$classifier").toNIO)
  }

  /**
   * Every module the dump covers. Test modules are excluded on purpose: uika checks what ships,
   * and a test-only dependency is never on the runtime classpath the check compares.
   */
  private def javaModules(ev: Evaluator): Seq[JavaModule] =
    ev.resolveModulesOrTasks(Seq("__"), SelectMode.Multi).get.collect {
      case Left(m: JavaModule) if !m.isInstanceOf[TestModule] => m
    }

  /**
   * One module's dump entry.
   *
   * `Task.traverse` rather than calling `dep.localRunClasspath()` in the body: the dependency list
   * is only known at runtime, and Mill's task macro can only lift statically known task calls
   * into edges.
   */
  private def moduleDumpTask(m: JavaModule): Task[ClasspathDump.Module] = {
    val depModules = m.recursiveRunModuleDeps
    val depOutputs = Task.traverse(depModules)(_.localRunClasspath)
    Task.Anon {
      val ownOutput = m.localRunClasspath().map(_.path)
      val classesDirs = ownOutput.filter(os.exists)

      // Coordinates come from the resolution, never from file paths. Mill models module deps as
      // synthetic coursier projects with no publications, so this yields external artifacts only
      // and the internal ones are attributed from `depOutputs` below.
      //
      // withConfiguration is deprecated in coursier 2.1.25, but JavaModule.resolvedRunMvnDeps
      // still builds the runtime dependency exactly this way, and matching it verbatim is the
      // point: the coordinates must describe the resolution the module actually runs on.
      val runtimeDep: cs.Dependency =
        m.coursierDependencyTask().withConfiguration(cs.Configuration.runtime): @nowarn(
          "cat=deprecation"
        )
      val resolved = m.millResolver().fetchArtifacts(Seq(BoundDep(runtimeDep, force = false)))
      // The RESOLVED version, not `dep.versionConstraint.asString`: a declared range or a
      // dynamic version would otherwise be written verbatim, and two dumps taken either side
      // of a real upgrade would carry the same constraint string and diff to no change.
      val resolvedVersions = resolved.resolution.projectCache0.map {
        case (key, (_, project)) => key -> project.version0.asString
      }
      val coordinates = resolved.fullDetailedArtifacts0.collect {
        case (dep, _, _, Some(file)) =>
          os.Path(file) -> (
            dep.module.organization.value,
            dep.module.name.value,
            resolvedVersions.getOrElse(
              (dep.module, dep.versionConstraint),
              dep.versionConstraint.asString
            )
          )
      }.toMap

      // Only a dep module's own OUTPUT carries its `project` label: the key tells uika it may
      // substitute that module's classesDirs when the file is missing, which is a lie for a
      // jar the dep merely puts on the classpath (`unmanagedClasspath`, `compileResources`).
      val projectOf = depModules.zip(depOutputs()).flatMap { case (dep, entries) =>
        entries.map(_.path -> moduleLabel(dep))
      }.toMap

      // Subtract exactly what this module produces, and NOT `localClasspath()` -- that also
      // holds `unmanagedClasspath()`, so subtracting it dropped a module's vendored jars from
      // the dump altogether even though they are on the classpath it runs on.
      // `os.exists` because a module names its resource directories whether or not they were
      // ever created, and an entry pointing at nothing is noise in every report.
      val own = (ownOutput ++ m.compileResources().map(_.path)).toSet
      val entries = m.runClasspath().map(_.path).distinct.filterNot(own).filter(os.exists)
      val artifacts = entries.map { path =>
        val (group, name, version) = coordinates.getOrElse(path, (null, null, null))
        new ClasspathDump.Artifact(group, name, version, path.toString, projectOf.get(path).orNull)
      }

      // What THIS module compiles for, in the dump next to it, so upgrade-check can scope a
      // JDK move to the modules that made it. mandatoryJavacOptions as well as javacOptions,
      // for the reason upgradeCheck spells out: Mill compiles with both.
      val declared = Seq(m.javacOptions(), m.mandatoryJavacOptions())
        .flatMap(options => Option(UikaCli.declaredRelease(options.asJava)))
      new ClasspathDump.Module(
        moduleLabel(m),
        classesDirs.map(_.toString).asJava,
        artifacts.asJava,
        if (declared.isEmpty) null else declared.minBy(_.intValue)
      )
    }
  }

  /** `:foo:bar`, the `:path` shape the dump format uses for Gradle and Maven modules too. */
  private def moduleLabel(m: JavaModule): String = ":" + m.moduleSegments.parts.mkString(":")

  lazy val millDiscover: Discover = Discover[this.type]
}

File: clojure-tool/src/exoego/uika.clj
(ns exoego.uika
  "uika as a Clojure CLI tool: `clojure -T:uika dump-classpath` / `clojure -T:uika upgrade-check`.

  Published to Maven Central as net.exoego.uika/clojure-uika and consumed through a
  deps.edn alias that carries :ns-default itself, because tools.deps resolves no
  usage data for :mvn coordinates (coord-usage :mvn is TBD upstream), so a
  `-Ttools`-installed Maven tool would need every function call qualified. The
  tool's own :mvn/version in the runtime basis names the matching uika-cli
  release, so the alias pins both. Everything not specific to tools.deps lives
  in exoego.uika.core, shared with the Leiningen plugin."
  (:require [clojure.java.basis :as basis]
            [clojure.java.io :as io]
            [clojure.tools.deps :as deps]
            [clojure.tools.deps.util.dir :as deps-dir]
            [exoego.uika.core :as core]))

(set! *warn-on-reflection* true)

(defn- project-basis
  "The PROJECT's basis, not this tool's own. `create-basis` re-resolves exactly the
  way `clojure -M`/`-X` would; the user config is excluded because the application
  ships without it. `:aliases` widens the resolution the same way the user's run
  alias would (e.g. `:aliases [:prod]`)."
  [dir {:keys [aliases]}]
  ;; with-dir rebinds the directory EVERY relative path resolves against -- the
  ;; :project file and the deps.edn's own :local/root entries alike -- so :project
  ;; stays the bare file name. A dir-joined :project would resolve twice.
  (deps-dir/with-dir (.getCanonicalFile (io/file dir))
    (deps/create-basis (cond-> {:user nil :project "deps.edn"}
                         (seq aliases) (assoc :aliases (mapv keyword aliases))))))

(defn dump-classpath
  "Writes the project's resolved classpath as a uika v2 dump.

  :output   file to write (default \"target/uika/classpath.json\")
  :dir      project directory (default: where the tool was invoked)
  :aliases  aliases to include in the resolution, e.g. [:prod]
  :class-dir extra classes directory to record (AOT output from tools.build
             compile-clj); the project :paths are always recorded so hinted-interop
             classes compiled there are scanned too.
  :jdk-release the release to record the application as running on, instead of this
             JVM's. The same knob upgrade-check takes, for a project whose runtime is
             not the JVM that writes the dump. 0 leaves the recorded release derived,
             since it only switches the API layer off."
  [{:keys [dir output class-dir jdk-release] :as args}]
  (let [dir (str (or dir (System/getProperty "user.dir")))
        basis (project-basis dir args)
        ;; Source dirs go in as classesDirs: uika only parses .class files, so pure
        ;; .clj trees contribute nothing, but AOT output or java compilation landing
        ;; in :paths is scanned without further configuration.
        class-dirs (cond-> (into [] (comp (map #(str (io/file dir %)))
                                          (filter #(.isDirectory (io/file ^String %))))
                                 (:paths basis))
                     class-dir (conj (str (io/file dir ^String class-dir))))
        ;; One lib can contribute several paths, so this is a nested for, not a map.
        artifacts (vec (for [[lib {:mvn/keys [version] :keys [paths]}] (sort-by key (:libs basis))
                             path paths]
                         (core/lib->artifact lib version path)))
        out (let [f (io/file (or output "target/uika/classpath.json"))]
              (if (.isAbsolute f) f (io/file dir (str f))))]
    (io/make-parents out)
    (spit out (core/dump-json (str ":" (.getName (io/file dir))) artifacts class-dirs
                              (or (core/override-release jdk-release)
                                  (.feature (Runtime/version)))))
    (println "uika classpath dump:" (str out))
    (str out)))

(defn version-from-libs
  "The version this tool was resolved with, out of a basis :libs map: its own
  Maven coordinate, the deps.edn alias flow the docs describe. A git or
  :local/root install carries no such coordinate and falls back to
  :cli-version / UIKA_CLI_VERSION, with the resolve-binary usage hint naming
  the former."
  [libs]
  (some (fn [[lib {:mvn/keys [version]}]]
          (when (= lib 'net.exoego.uika/clojure-uika) version))
        libs))

(defn- own-version
  []
  (version-from-libs (:libs (basis/current-basis))))

(defn upgrade-check
  "Runs uika upgrade-check over a before/after pair of dumps.

  :before / :after   dump files (required)
  :fail-on           never | reachable | any (CLI default when omitted)
  :exclude-file      one path or a vector of paths
  :jdk-release       JDK API release; 0 disables, default is the running JVM's
  :jfr               runtime class-load evidence as JFR: a recording, or a directory
                     holding recordings and text logs mixed. Recordings are converted
                     with the JDK's own JFR reader before the CLI runs, which needs
                     this tool on Java 17+
  :class-load-log    one path or a vector: runtime class-load evidence in the CLI's
                     text format (a recording given here is converted too)
  :draft-exclude-file
                     where the CLI writes draft exclude rules for symbols never
                     observed loading, which needs :class-load-log too
  :cli-version       uika-cli version; defaults to this tool's own version, read
                     from its Maven coordinate in the runtime basis
                     (UIKA_CLI_VERSION is consulted in between, and a source
                     install has no coordinate, so it needs one of the two)
  :cli-path          existing uika binary, skipping the download entirely
                     (UIKA_CLI_PATH does the same from the environment)"
  [{:keys [before after] :as args}]
  (when-not (and before after)
    (throw (ex-info "usage: clojure -Tuika upgrade-check :before <a.json> :after <b.json>" {})))
  (core/run-upgrade-check
   (core/resolve-binary args own-version "uika-cli version is unknown; pass :cli-version")
   args))
